src/rdkafka_queue.c

```diff
@@ -72,16 +72,19 @@ rd_kafka_q_t *rd_kafka_q_new0 (rd_kafka_t *rk, const char *func, int line) {
  * All access to rkq_fwdq are protected by rkq_lock.
  */
 void rd_kafka_q_fwd_set0 (rd_kafka_q_t *srcq, rd_kafka_q_t *destq,
-                          int do_lock) {
+                          int do_lock, int fwd_app) {
 
         if (do_lock)
                 mtx_lock(&srcq->rkq_lock);
+        if (fwd_app)
+                srcq->rkq_flags |= RD_KAFKA_Q_F_FWD_APP;
 	if (srcq->rkq_fwdq) {
 		rd_kafka_q_destroy(srcq->rkq_fwdq);
 		srcq->rkq_fwdq = NULL;
 	}
 	if (destq) {
-		rd_kafka_q_keep(destq);
+                rd_kafka_q_lock(destq);
```

Comment:
This isnt safe, q_concat() will lock destq again and cause a deadlock.
What's the problem with the original code?